agora/Agora/Utils.hs

```diff
@@ -76,24 +81,30 @@ passert :: Term s PString -> Term s PBool -> Term s k -> Term s k
 passert errorMessage check k = pif check k (ptraceError errorMessage)
 
 -- | Find a datum with the given hash.
-pfindDatum :: Term s (PDatumHash :--> PTxInfo :--> PMaybe PDatum)
+pfindDatum :: Term s (PDatumHash :--> PBuiltinList (PAsData (PTuple PDatumHash PDatum)) :--> PMaybe PDatum)
 pfindDatum = phoistAcyclic $
-  plam $ \datumHash txInfo'' -> P.do
-    PTxInfo txInfo' <- pmatch txInfo''
-    plookupTuple # datumHash #$ pfield @"data" # txInfo'
+  plam $ \datumHash datums -> plookupTuple # datumHash # datums
+
+-- | Find a datum with the given hash, and `ptryFrom` it.
+ptryFindDatum :: forall (a :: PType) (s :: S). PTryFrom PData a => Term s (PDatumHash :--> PBuiltinList (PAsData (PTuple PDatumHash PDatum)) :--> PMaybe a)
+ptryFindDatum = phoistAcyclic $
+  plam $ \datumHash inputs -> P.do
+    pmatch (pfindDatum # datumHash # inputs) $ \case
+      PNothing -> pcon PNothing
+      PJust datum -> P.do
+        (datum', _) <- ptryFrom (pto datum)
+        pcon (PJust datum')
 
 {- | Find a datum with the given hash.
 NOTE: this is unsafe in the sense that, if the data layout is wrong, this is UB.
```

Comment:
UB?

Reply:
Yeah, I'm not sure if this undefined behaviour or not either.

Reply:
It did mean undefined behavior, but we can remove this function now.